[WIP] Actors quickstart doc (#3309)
dapr/docs: daprdocs/content/en/getting-started/quickstarts/actors-quickstart.md

---
type: docs
title: "Quickstart: Actors"
linkTitle: "Actors"
weight: 75
description: "Get started with Dapr's Actors building block"
---

Let's take a look at Dapr's [Actors building block]({{< ref actors >}}). In this Quickstart, you will run a `SmartDevice.Service` microservice and a simple console client to demonstrate the stateful object patterns in Dapr Actors.  
1. Using a `SmartDevice.Service` microservice, you can host:
   - Two `SmartDectectorActor` smoke alarm objects
   - A `ControllerActor` object that commands and controls the smart devices  
1. Using a `SmartDevice.Client` console app, the client app interacts with each actor, or the controller, to perform actions in aggregate. 
1. The `SmartDevice.Interfaces` contains the shared interfaces and data types used by both the service and client apps.

> <img src="/images/bindings-quickstart/bindings-quickstart.png" width=800 style="padding-bottom:15px;">

Reviewer: Need a new diagram drawn here which shows 2 device actors and one controller actor. On each actor show the method endpoint and then shows the calls between the actors with numbers. These numbers should follow the same call sequence as describe in the "When you ran the client app:" section below.

Select your preferred language-specific Dapr SDK before proceeding with the Quickstart.

{{< tabs ".NET" >}}
 <!-- .NET -->
{{% codetab %}}

### Pre-requisites

For this example, you will need:

- [Dapr CLI and initialized environment](https://docs.dapr.io/getting-started).
- [.NET SDK or .NET 6 SDK installed](https://dotnet.microsoft.com/download).
<!-- IGNORE_LINKS -->
- [Docker Desktop](https://www.docker.com/products/docker-desktop)
<!-- END_IGNORE -->

### Step 1: Set up the environment

Clone the [sample provided in the Quickstarts repo](https://github.com/dapr/quickstarts/tree/master/bindings).

```bash
git clone https://github.com/dapr/quickstarts.git
```

### Step 2: Run the service app

In a new terminal window, navigate to the `actors/csharp/sdk/service` directory and restore dependencies:

```bash
cd actors/csharp/sdk/service
dotnet build
```

Run the `SmartDevice.Service`, which will start service itself and the Dapr sidecar:

```bash

> dapr run --app-id actorservice --app-port 5001 --dapr-http-port 3500 --components-path ../../../resources -- dotnet run --urls=http://localhost:5001/

Reviewer: ```suggestion
dapr run --app-id actorservice --app-port 5001 --dapr-http-port 3500 --resources-path ../../../resources -- dotnet run --urls=http://localhost:5001/
```

        // Create a new instance of the data class that will be stored in the actor
        var deviceData1 = new SmartDeviceData(){
            Location = "First Floor",
            Status = "Ready",
        };
        // Now you can use the actor interface to call the actor's methods.
        Console.WriteLine($"Calling SetDataAsync on {smokeDetectorActorType}:{deviceActorId1}...");
        var setDataResponse1 = await proxySmartDevice1.SetDataAsync(deviceData1);
        Console.WriteLine($"Got response: {setDataResponse1}");
```

The `ControllerActor` object is used to keep track of the devices and trigger the alarm for all of them.

```csharp
        var controllerActorId = new ActorId("controller");
        var proxyController = ActorProxy.Create<IController>(controllerActorId, controllerActorType);
        Console.WriteLine($"Registering the IDs of both Devices...");
        await proxyController.RegisterDeviceIdsAsync(new string[]{deviceId1, deviceId2});
        var deviceIds = await proxyController.ListRegisteredDeviceIdsAsync();
        Console.WriteLine($"Registered devices: {string.Join(", " , deviceIds)}");
```

Additionally look at:

> - `SmartDevice.Service/SmartDetectorActor.cs` which contains the implementation of the the smart device actor actions

Reviewer: ```suggestion
- `SmartDevice.Service/SmartDetectorActor.cs` which contains the implementation of the the smart device actor
```

- `SmartDevice.Service/ControllerActor.cs` which contains the implementation of the controller actor that manages all devices

> - `SmartDevice.Interfaces/ISmartDevice` which contains the required actions and shared data types for each SmartDetectorActor

Reviewer: ```suggestion
- `SmartDevice.Interfaces/ISmartDevice` which contains the method definitions and shared data types for each SmartDetectorActor
```

> - `SmartDevice.Interfaces/IController` which contains the actions a controller can perform across all devices

Reviewer: ```suggestion
- `SmartDevice.Interfaces/IController` which contains the method definitions and shared data types for the ControllerActor
```

== APP == info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
== APP ==       Request finished HTTP/1.1 GET http://127.0.0.1:5001/healthz - - - 200 - text/plain 5.2599ms
```

### Step 3: Run the client app

In a new terminal instance, navigate to the `actors/csharp/sdk/client` directory and install the dependencies:

```bash
cd ./actors/csharp/sdk/client
dotnet build
```

Run the `SmartDevice.Client` app:

```bash
dapr run --app-id actorclient -- dotnet run
```

Expected output:

```bash
== APP == Startup up...
== APP == Calling SetDataAsync on SmokeDetectorActor:1...
== APP == Got response: Success
== APP == Calling GetDataAsync on SmokeDetectorActor:1...
== APP == Device 1 state: Location: First Floor, Status: Ready
== APP == Calling SetDataAsync on SmokeDetectorActor:2...
== APP == Got response: Success
== APP == Calling GetDataAsync on SmokeDetectorActor:2...
== APP == Device 2 state: Location: Second Floor, Status: Ready
== APP == Registering the IDs of both Devices...
== APP == Registered devices: 1, 2
== APP == Detecting smoke on Device 1...
== APP == Device 1 state: Location: First Floor, Status: Alarm
== APP == Device 2 state: Location: Second Floor, Status: Alarm

> ```

Reviewer: Add these update based on this PR 
https://github.com/dapr/quickstarts/pull/838/files

```suggestion
== APP == Sleeping for 16 seconds before checking status again to see reminders fire and clear alarms
== APP == Device 1 state: Location: First Floor, Status: Ready
== APP == Device 2 state: Location: Second Floor, Status: Ready
```
```
